Ndev 1834: Heap inside Account

evgenyzdanovich opened the pull request:
First revision of the change to have initial code suggestions. Several things yet to finish and improve.

-PersistentState contains two tightly connected structs - EvmMachine and ExecutorState. EvmMachine::execute takes ExecutorState as a mutable borrow and having both in one struct (PersistentState) is not compatible with borrow checker.
-To perform the same changes for the transaction_execute (additionally introduce the holder account for dependent instructions).

evgenyzdanovich committed:
Second heap on top of holder account - first revision. Currently, only for the transaction_step (transaction_execute in the next commits).

andreisilviudragnea commented on evm_loader/program/src/account/state.rs:
```diff
@@ -1,13 +1,15 @@
+use std::ptr::addr_of_mut;
 
-use crate::account_storage::AccountStorage;
+use crate::account_storage::{AccountStorage, ProgramAccountStorage};
```
I suggest using conditional compilation with `#[cfg(target_os = "solana")]` for code used only in NeonEVM Solana program, until at least the feature of this PR is implemented. Then the code can be decoupled more to reduce the usages of `#[cfg(target_os = "solana")]`.

```Rust
use crate::account_storage::AccountStorage;
#[cfg(target_os = "solana")]
use crate::account_storage::ProgramAccountStorage;
```

evgenyzdanovich replied:
Thanks for the feedback! As discussed on the meeting, I went with a different path: split ExecutorState into persistent part (struct with all data owned, now called ExecutorStateData) and into refs (old Executor State, mutably borrows ExecutorState).

andreisilviudragnea commented on evm_loader/program/src/account/state.rs:
```diff
@@ -98,37 +103,89 @@ impl<'a> StateAccount<'a> {
         })
     }
 
+    pub fn alloc<'r>(&self, account_storage: &'r ProgramAccountStorage<'a>) -> Result<()> {
```
This method and all newly added ones need `#[cfg(target_os = "solana")]`. The compilation errors should guide the usage of `#[cfg(target_os = "solana")]`.

andreisilviudragnea reviewed (commented):
I suggest first focusing on getting `cargo build-sbf` and `cargo build` compile successfully.
I also suggest using `cargo fmt` and `cargo clippy` sooner than later, since both commands have many suggestions for the current code.

evgenyzdanovich committed:
Factor out program specific functions of StateAccount under conditional compilation section.

evgenyzdanovich committed:
Split StateAccountAllocator for Solana/non-Solana case.

evgenyzdanovich committed:
Adjust EVM Machine to use Box with account allocator.

evgenyzdanovich committed:
Simplify the StateAccount loading/storing of persistent objects for the iterative transactions.

evgenyzdanovich committed:
Adjust emulation path according to changed ExecutorState interface.

evgenyzdanovich committed:
Fix cargo clippy and apply cargo fmt.

evgenyzdanovich committed:
Fix the test.

evgenyzdanovich committed:
Add holder account into transaction_execute instructions.

evgenyzdanovich committed:
Fix double-borrow of state account data.

evgenyzdanovich committed:
Fix double dealloc, incorrect box leak and double borrow.

evgenyzdanovich committed:
Temporary adjust heap offset and initialization. Add repr C and move some more data structs to use persistent data structures.

evgenyzdanovich committed:
Fix double-free error caused by the allocator-api2's Box.

evgenyzdanovich committed:
Box the Data of StateAccount to persist it between iterations. Previously, the data was instantiated on the stack hence not persisted.

evgenyzdanovich committed:
cargo fmt && cargo clippy and cargo build (non-sbf) compication fixes.

neonlabstech commented:
<details><summary>Dapps report</summary>

evgenyzdanovich committed:
Make allocator to rely on dynamic offset of the heap object. Init heap object in the state and in holder differently. Take buffer of the Holder (e.g. Transaction struct) when calculating heap offset.

evgenyzdanovich committed:
Fixes after rebase.

neonlabstech commented:
<details><summary>Dapps report</summary>

neonlabstech commented:
<details><summary>Dapps report</summary>

neonlabstech commented:
<details><summary>Dapps report</summary>

evgenyzdanovich committed:
Introduce the method for neon-core-api to read bits of data from state account.

neonlabstech commented:
<details><summary>Dapps report</summary>

neonlabstech commented:
<details><summary>Dapps report</summary>

evgenyzdanovich committed:
Adjust heap initialization in transactions from account: copy transaction rlp, init heap, reuse transaction rlp bytes.

neonlabstech commented:
<details><summary>Dapps report</summary>

neonlabstech commented:
<details><summary>Dapps report</summary>

neonlabstech commented:
<details><summary>Dapps report</summary>

neonlabstech commented:
<details><summary>Dapps report</summary>

neonlabstech commented:
<details><summary>Dapps report</summary>

neonlabstech commented:
<details><summary>Dapps report</summary>

evgenyzdanovich committed:
Tiny compilation fix after the sync with develop.

neonlabstech commented:
<details><summary>Dapps report</summary>

evgenyzdanovich committed:
Remove Buffer's serialization Deserialization impls and its corresponding tests.

evgenyzdanovich committed:
Remove Buffer's Account Uninit that was used in he serialization case. Buffers after that are reinitialized by matching to Account variant and applying the data pointer + offset.

alfiedotwtf commented on evm_loader/program/src/account/mod.rs:
```diff
@@ -188,6 +193,59 @@ pub fn revision(program_id: &Pubkey, info: &AccountInfo) -> Result<u32> {
+    let heap_ptr = heap_ptr.cast_mut();
+    unsafe {
+        // First, zero out underlying bytes of the future heap representation.
+        heap_ptr.write_bytes(0, size_of::<Heap>());
```
Is it worth using https://docs.rs/zeroize/latest/zeroize/ for something like this (and in `delete()`)

evgenyzdanovich replied:
thanks for suggestion! I'm a bit hesitant to pull up additional library just for this one case. Are there other cases where we zero memory?

alfiedotwtf replied:
All good, feel free to ignore. It was only a suggestion and something to keep in mind in the future.

neonlabstech commented:
<details><summary>Dapps report</summary>


Cost report for "Aave" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Token mint | 0.00234464 | 0.0037536141985631995 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| Token approve | 0.00234464 | 0.0037536141985631995 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| Deposit to lending pool | 0.00796264 | 0.012747662220529601 | 4 | 5 | 4845360 | 3981320 | 82.17 | 
| Borrow from lending pool | 0.00519928 | 0.008323547109264 | 4 | 30 | 3793040 | 2599640 | 68.54 | 
| Repay | 0.00049 | 0.000784442862 | 4 | 27 | 2586800 | 245000 | 9.47 | 
| Flashloan | 0.00512928 | 0.008211483843264 | 4 | 29 | 7499680 | 2564640 | 34.2 | 
| Withdraw | 7e-05 | 0.000112063266 | 4 | 5 | 1857760 | 35000 | 1.88 | 
| Liquidation | 0.00341464 | 0.005466510151632 | 4 | 57 | 4967840 | 1707320 | 34.37 | 

Cost report for "Compound-Finance" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| borrow erc20 cToken | 0.00512864 | 0.008210459264832 | 17 | 1 | 4154304 | 2564320 | 61.73 | 
| borrow ether cToken | 0.002804 | 0.00448889586844 | 14 | 1 | 2601104 | 1402000 | 53.9 | 
| Repay borrow erc20 cToken | 2e-05 | 3.2017802200000006e-05 | 18 | 1 | 1362240 | 10000 | 0.73 | 
| Mint erc20 cToken | 2e-05 | 3.2017802200000006e-05 | 17 | 1 | 1087592 | 10000 | 0.92 | 
| Mint ether cToken | 2e-05 | 3.2017802200000006e-05 | 12 | 1 | 296648 | 10000 | 3.37 | 
| Transfer erc20 cToken | 0.0116432 | 0.018639483728752 | 22 | 1 | 6788408 | 5821600 | 85.76 | 
| Add reserves for ether cToken | 2e-05 | 3.2017802200000006e-05 | 10 | 1 | 27500 | 10000 | 36.36 | 
| Liquidate borrow | 2e-05 | 3.2017802200000006e-05 | 21 | 1 | 1933536 | 10000 | 0.52 | 
| Redeem erc20 cToken | 2e-05 | 3.2017802200000006e-05 | 17 | 1 | 823944 | 10000 | 1.21 | 

Cost report for "Robonomics" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| PublicAmbix setSink | 0.00466928 | 0.0074750041728208 | 11 | 1 | 199023640 | 2334640 | 1.17 | 
| PublicAmbix appendSource | 0.00931856 | 0.014917990543441602 | 13 | 1 | 199023640 | 4659280 | 2.34 | 
| XRT approve | 0.00234464 | 0.0037535109875104 | 10 | 1 | 199022872 | 1172320 | 0.59 | 
| PublicAmbix run | 2e-05 | 3.2017802200000006e-05 | 21 | 1 | 199021400 | 10000 | 0.01 | 
| KycAmbix setSink | 0.00466928 | 0.0074750041728208 | 11 | 1 | 199023576 | 2334640 | 1.17 | 
| KycAmbix appendSource | 0.00931856 | 0.014917990543441602 | 13 | 1 | 199023576 | 4659280 | 2.34 | 
| KycAmbix run | 2e-05 | 3.2017802200000006e-05 | 20 | 1 | 199021400 | 10000 | 0.01 | 
| DutchAuction started | 2e-05 | 3.2017802200000006e-05 | 9 | 1 | 199021272 | 10000 | 0.01 | 
| DutchAuction bid | 0.00234464 | 0.0037535109875104 | 14 | 1 | 199026200 | 1172320 | 0.59 | 
| DutchAuction update stage | 2e-05 | 3.2017802200000006e-05 | 9 | 1 | 199021272 | 10000 | 0.01 | 
| DutchAuction claim tokens | 2e-05 | 3.2017802200000006e-05 | 13 | 1 | 199021272 | 10000 | 0.01 | 
| XRT transfer | 0.00234464 | 0.0037535109875104 | 11 | 1 | 199023192 | 1172320 | 0.59 | 
| Factory create lighthouse | 0.0417104 | 0.066773766844144 | 19 | 1 | 199022488 | 20855200 | 10.48 | 
| XRT increase allowance | 0.00234464 | 0.0037535109875104 | 10 | 1 | 199022936 | 1172320 | 0.59 | 

Cost report for "Yearn" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Create new release | 0.00234464 | 0.0037535430856319996 | 14 | 1 | 1322552 | 1172320 | 88.64 | 
| Clone strategy | 0.03887072 | 0.062225534808124795 | 16 | 1 | 21455896 | 19435360 | 90.58 | 
| Set strategist authority | 2e-05 | 3.2018076000000005e-05 | 9 | 1 | 27500 | 10000 | 36.36 | 
| Set emergency shutdown | 2e-05 | 3.2018076000000005e-05 | 9 | 1 | 27500 | 10000 | 36.36 | 
| Emergency shutdown | 2e-05 | 3.2018076000000005e-05 | 9 | 1 | 27500 | 10000 | 36.36 | 
| Call strategy harvest | 0.00699392 | 0.011196593104896 | 23 | 1 | 7854000 | 3496960 | 44.52 | 
| Withdraw from vault | 2e-05 | 3.2018076000000005e-05 | 15 | 1 | 823944 | 10000 | 1.21 | 

Cost report for "Uniswap V3" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Token approve | 0.00234464 | 0.0037535430856319996 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| NonfungiblePositionManager - Mint position | 0.02786064 | 0.0446032637079648 | 4 | 28 | 31435120 | 13930320 | 44.31 | 
| NonfungiblePositionManager - Increase liquidity  | 0.00044 | 0.0007044144008000001 | 23 | 23 | 4364240 | 220000 | 5.04 | 
| NonfungiblePositionManager - Decrease Liquidity | 0.00139808 | 0.0022382447397056 | 18 | 1 | 4274240 | 699040 | 16.35 | 
| NonfungiblePositionManager - Collect Fees | 0.00466928 | 0.0074752456212896 | 21 | 1 | 3553040 | 2334640 | 65.71 | 
| NonfungiblePositionManager - Burn Liquidity Position | 2e-05 | 3.2018836400000004e-05 | 16 | 1 | 3954560 | 10000 | 0.25 | 
| Direct swap | 2e-05 | 3.2018836400000004e-05 | 18 | 1 | 1328400 | 10000 | 0.75 | 
| Burn transaction | 0.00047936 | 0.0007674274708352 | 12 | 1 | 739040 | 239680 | 32.43 | 
| Collect transaction | 0.00234464 | 0.0037536322288447997 | 13 | 1 | 1661680 | 1172320 | 70.55 | 

Cost report for "Saddle Finance" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Add liquidity | 0.00235464 | 0.0037696416470448 | 3 | 2 | 3209760 | 1177320 | 36.68 | 
| Swap DAI -> USDC | 2e-05 | 3.2018836400000004e-05 | 22 | 1 | 1797760 | 10000 | 0.56 | 
| Swap USDC -> DAI | 2e-05 | 3.2018836400000004e-05 | 22 | 1 | 1797760 | 10000 | 0.56 | 
| Remove liquidity | 2e-05 | 3.2018836400000004e-05 | 24 | 1 | 1937440 | 10000 | 0.52 | 
| Add liquidity in 3 Tokens | 0.00284464 | 0.00455410791784 | 4 | 31 | 4248480 | 1422320 | 33.48 | 
| Swap DAI -> USDC 3 pool | 2e-05 | 3.2018870000000004e-05 | 22 | 1 | 1807760 | 10000 | 0.55 | 
| Swap USDC -> USDT 3 pool | 2e-05 | 3.2018802000000005e-05 | 22 | 1 | 1807760 | 10000 | 0.55 | 
| Swap USDC -> DAI 3 pool | 2e-05 | 3.2018349800000005e-05 | 22 | 1 | 1807760 | 10000 | 0.55 | 
| Remove liquidity 3 pool | 3e-05 | 4.80275247e-05 | 3 | 2 | 2646480 | 15000 | 0.57 | 
| Add liquidity in metapool | 0.00291464 | 0.0046660981530536 | 3 | 31 | 3299760 | 1457320 | 44.16 | 
| Swap SUSD -> LP metapool | 2e-05 | 3.2018349800000005e-05 | 23 | 1 | 1807760 | 10000 | 0.55 | 
| Swap USDC -> SUSD metapool | 0.00103 | 0.0016489658722000002 | 4 | 54 | 3775520 | 515000 | 13.64 | 
| Swap DAI -> USDT metapool | 0.00056 | 0.0008965216623999999 | 4 | 30 | 3086160 | 280000 | 9.07 | 
| Remove liquidity in Metapool | 2e-05 | 3.20186308e-05 | 24 | 1 | 1937440 | 10000 | 0.52 | 

Cost report for "Uniswap V2" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Token approve | 0.00234464 | 0.0037536081259455997 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| Direct swap | 2e-05 | 3.2018870000000004e-05 | 19 | 1 | 1048720 | 10000 | 0.95 | 

Cost report for "Curve" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Metapool: add liquidity | 0.01664256 | 0.0266436885808896 | 22 | 229 | 12479456 | 8321280 | 66.68 | 
| Metapool: swap 1 | 0.00458 | 0.00733232123 | 19 | 230 | 4803832 | 2290000 | 47.67 | 
| Metapool: swap 2 | 0.00464 | 0.00742837784 | 19 | 233 | 4836832 | 2320000 | 47.97 | 
| Metapool: withdraw admin fees | 2e-05 | 3.2018831800000004e-05 | 16 | 1 | 549296 | 10000 | 1.82 | 
| Pool: add liquidity | 0.01384464 | 0.0221644626050592 | 26 | 577 | 11162624 | 6922320 | 62.01 | 
| Pool: swap 1 | 0.00542 | 0.0086769727958 | 19 | 272 | 5265832 | 2710000 | 51.46 | 
| Pool: swap 2 | 0.00542 | 0.0086769727958 | 19 | 272 | 5265832 | 2710000 | 51.46 | 
| Pool: swap 3 | 0.00542 | 0.008677104664400002 | 19 | 272 | 5265832 | 2710000 | 51.46 | 
| Pool: withdraw admin fees | 2e-05 | 3.2018836400000004e-05 | 17 | 1 | 55000 | 10000 | 18.18 | 
| Stablecoin: transfer | 2e-05 | 3.2018836400000004e-05 | 11 | 1 | 538296 | 10000 | 1.86 | 
| Tri pool: add liquidity | 0.01561656 | 0.0250012039885392 | 3 | 39 | 13440856 | 7808280 | 58.09 | 
| Tri pool: swap 1 | 0.00072 | 0.0011526781104000002 | 20 | 37 | 2680832 | 360000 | 13.43 | 
| Tri pool: swap 2 | 0.00072 | 0.0011526781104000002 | 20 | 37 | 2680832 | 360000 | 13.43 | 
| Tri pool: swap 3 | 0.0007 | 0.001120655697 | 20 | 36 | 2669832 | 350000 | 13.11 | 
| Tri pool: apply new fee | 2e-05 | 3.20186892e-05 | 9 | 1 | 27500 | 10000 | 36.36 | 
| Tri pool: withdraw admin fees | 2e-05 | 3.20187826e-05 | 19 | 1 | 560296 | 10000 | 1.78 | 
| gauge v3: deposit | 0.01907648 | 0.0305402832946624 | 29 | 1 | 11085360 | 9538240 | 86.04 | 
| gauge controller: vote | 0.01812992 | 0.0290249471212544 | 27 | 1 | 14547720 | 9064960 | 62.31 | 
| gauge controller: vote | 0.01069664 | 0.0171245591461984 | 27 | 1 | 10444104 | 5348320 | 51.21 | 
| gauge controller: vote | 0.01069664 | 0.0171246983094848 | 27 | 1 | 10444104 | 5348320 | 51.21 | 
| gauge controller: vote | 0.00791264 | 0.012667676282604802 | 27 | 1 | 8392296 | 3956320 | 47.14 | 
| gauge controller: vote | 0.005588 | 0.00894606289016 | 27 | 1 | 7366392 | 2794000 | 37.93 | 
| gauge controller: vote | 0.005588 | 0.00894606289016 | 27 | 1 | 7366392 | 2794000 | 37.93 | 
| gauge controller: vote | 0.00791264 | 0.012667676282604802 | 27 | 1 | 8392296 | 3956320 | 47.14 | 
| gauge controller: vote | 0.005588 | 0.00894606289016 | 27 | 1 | 7366392 | 2794000 | 37.93 | 
| gauge controller: vote | 0.005588 | 0.00894606289016 | 27 | 1 | 7366392 | 2794000 | 37.93 |

alfiedotwtf commented on evm_loader/program/src/account/state.rs:
```diff
@@ -22,68 +31,179 @@ pub enum AccountsStatus {
+                .cast_mut();
+            let account_slice = slice::from_raw_parts(accounts_buf_ptr_adjusted, accounts_len);
+            // Allocate a new vector and with the exact number of elements and copy the memory.
+            let mut accounts = vec![Pubkey::default(); accounts_len];
```
What's our `vector` Victor?
  -- Airplane!, 1980

There's a handful of `vec!` left around the place.

evgenyzdanovich replied:
This function `get_state_account_view` is intended to be used by the core-api. It has to return "revision" accounts (as was recently implemented by Anton) so this account list can be returned by the core-api (get_holder call).
So, short answer - there's no reason to use persistent vector. 

Let me also explain what's going on here, since it may be confusing.
Instead, we have to determine what's the offset from the start of account data where this account list of allocated. Because the state has all the data allocated within account, and the allocation is done in Solana runtime, all the pointers (**including buffers of persistent vectors**) are within that memory space (starts from 0x400000060).

When we are trying to read account data in the context of core-api, we are using system allocator (with standard, system dependent heap) which is different from aforementioned memory space (for example, on my linux machine it starts with 0x90000000 if I recall correctly). That's why we need careful low-level tricks to jump at the pointers and calculating the offset (variable `memory_space_delta` in this function) between the memory space that was used during allocation and current memory space. Hope it helps to navigate through, perhaps, the most confusing part of this PR. 

P.S. if you have an idea how to achieve the described goal (squeezing bits of information from the state/holder account) in a cleaner fashion - please shout out, I'm definitely open to suggestions.

alfiedotwtf replied:
Ah ok got it. I was under the impression ALL `vec` calls were to be replaced. All good.

alfiedotwtf commented on evm_loader/program/src/account/state.rs:
```diff
@@ -170,21 +304,9 @@ impl<'a> StateAccount<'a> {
     #[must_use]
     pub fn buffer(&self) -> Ref<[u8]> {
-        let data = self.account.data.borrow();
+        let data = self.account.try_borrow_data().unwrap();
```
Feel free to ignore, but do we have a policy on `unwrap()` vs `unwrap_X()` vs `expect()`?

evgenyzdanovich replied:
Leaving the decision to Anton, I'm also curious to learn what are the guidelines on how we agree to use unwrap/expect.

alfiedotwtf commented on evm_loader/program/src/account/state.rs:
```diff
@@ -311,4 +402,115 @@ impl<'a> StateAccount<'a> {
+    /// After this, the persistent objects can be allocated in the account data.
+    ///
+    /// N.B. No ownership checks are performed, it's a caller's responsibility.
+    /// TODO: This piece of should be moved to mod.rs.
```
These functions are now in `mod.rs` so they can be deleted (there's a slight difference though)

evgenyzdanovich replied:
correct, done

andreisilviudragnea committed:
Remove unused serde module

alfiedotwtf commented on evm_loader/program/src/account/state.rs:
```diff
@@ -311,4 +402,115 @@ impl<'a> StateAccount<'a> {
+    pub fn init_heap(info: &AccountInfo<'a>) -> Result<()> {
+        let (heap_ptr, heap_object_offset) = {
+            // Locate heap object after Holder's header.
+            let mut heap_object_offset = 100;
```
Is this always going to be 100 or should it be moved to a constant?

evgenyzdanovich replied:
this fn is unused, it was moved to mod.ts. Thanks for pointing out, I removed it.

alfiedotwtf commented on evm_loader/program/src/account_storage/apply.rs:
```diff
@@ -146,8 +146,8 @@ impl<'a> ProgramAccountStorage<'a> {
-                        data,
+                        accounts: accounts.to_vec(),
+                        data: data.to_vec(),
                     };
```
Should these also be wrapped in `to_vector`?

evgenyzdanovich replied:
I dont think so, this is a native Solana Instruction with a standard `vec` type.

alfiedotwtf commented on evm_loader/program/src/account/state.rs:
```diff
@@ -22,68 +31,179 @@ pub enum AccountsStatus {
+    pub revisions: TreeMap<Pubkey, u32>,
     /// Ethereum transaction gas used and paid
-    #[serde(with = "ethnum::serde::bytes::le")]
     pub gas_used: U256,
```
`bytes::le` has been removed a number of places now that we're using `repr(C)`. Even though Solana is LE, will this always guarantee the same endianess? I'm thinking even across different operators resuming an iteration from another operator.

evgenyzdanovich replied:
hm, great question. 
I think we do not rely on Solana being BE or LE. We rely on the persistence of the account's data and the fact that our SBF program is compiled, baked-in on-chain and ultimately shared as-is between operators. Does it make sense or I missed something?

alfiedotwtf replied:
Yep, makes sense, but that "shared as-is between operators" is what I'm not sure of. Maybe @anton-lisanin, @andreisilviudragnea, or @s-medvedev can comment.

alfiedotwtf commented on evm_loader/program/src/executor/state.rs:
```diff
@@ -123,21 +120,25 @@ impl<'a, B: AccountStorage> ExecutorState<'a, B> {
-        let mut accounts = BTreeMap::<Pubkey, OwnedAccountInfo>::new();
+        // Create accounts map on the default heap, as it's not part of the persistent state.
+        let mut accounts = std::collections::BTreeMap::<Pubkey, OwnedAccountInfo>::new();
 
```
Just checking, this is `BTreeMap` and not `TreeMap`.

evgenyzdanovich replied:
It's not part of persistent state, if I correctly understand the logic around. So yes, the standard BTreeMap is suitable here, as we dont want to flood account's persistent heap with unnecessary temporary object.

alfiedotwtf commented on evm_loader/program/src/executor/state.rs:
```diff
@@ -177,7 +178,7 @@ impl<'a, B: AccountStorage> ExecutorState<'a, B> {
 
-        Ok(accounts[&address].clone())
+        Ok(accounts.get(&address).unwrap().clone())
     }
```
Should this be an `expect()` instead?

evgenyzdanovich replied:
it can be, I feel it's related to adjacent comment on unwrap vs expect. Leaving unresolved and waiting for guidance from Anton and Semen.

alfiedotwtf commented on evm_loader/program/src/instruction/transaction_step.rs:
```diff
@@ -61,41 +60,67 @@ pub fn do_continue<'a>(
-    };
+        let mut state_data = storage.read_executor_state();
+        let mut evm = storage.read_evm();
+        //let (mut state, mut evm): (RefMut<'_, ExecutorStateData>, RefMut<'_, Evm>) = (storage.state(), storage.evm());
```
Should this line be deleted, or is it left as a comment?

evgenyzdanovich replied:
done

alfiedotwtf commented on evm_loader/program/src/instruction/transaction_step.rs:
```diff
@@ -61,41 +60,67 @@ pub fn do_continue<'a>(
+            &mut evm_backend,
+            None,
+        )?);
+        //storage.alloc(state_data, evm)?;
```
This comment too.

evgenyzdanovich replied:
done

andreisilviudragnea committed:
Refactor allocator module

Decouple Solana-only code - step 1

Less #[cfg(target_os = "solana")] and #[cfg(not(target_os = "solana"))]

Extract solana_allocator.rs module

Extract state_account_allocator.rs

Cleanup imports

Add useful comments

Promote allocator module to directory

Move solana_allocator.rs and state_account_allocator.rs into allocator.rs

Extract Alloc trait

Use acc_allocator() and make STATE_ACCOUNT_ALLOCATOR private

Remove static mut STATE_ACCOUNT_ALLOCATOR

Extract PROGRAM_DATA_INPUT_PARAMETERS_OFFSET constant

Extract impl_global_alloc macro

Extract solana module

Extract Alloc::heap method

Inline heap methods

evgenyzdanovich committed:
Make sure holder account is owned by the operator in TransactionExecuteFromInstruction.

evgenyzdanovich committed:
Code review changes, part 1.

evgenyzdanovich committed:
Code review changes, part 2.

neonlabstech commented:
<details><summary>Dapps report</summary>


Cost report for "Aave" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Token mint | 0.00234464 | 0.0035609219765536 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| Token approve | 0.00234464 | 0.0035609219765536 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| Deposit to lending pool | 0.00796264 | 0.0120932594203736 | 4 | 5 | 4845360 | 3981320 | 82.17 | 
| Borrow from lending pool | 0.00519928 | 0.0078964064480072 | 4 | 30 | 3793040 | 2599640 | 68.54 | 
| Repay | 0.00049 | 0.0007441874951000001 | 4 | 27 | 2586800 | 245000 | 9.47 | 
| Flashloan | 0.00512928 | 0.0077900939487072 | 4 | 29 | 7499680 | 2564640 | 34.2 | 
| Withdraw | 7e-05 | 0.0001063124993 | 4 | 5 | 1857760 | 35000 | 1.88 | 
| Liquidation | 0.00341464 | 0.005185984465853601 | 4 | 57 | 4967840 | 1707320 | 34.37 | 

Cost report for "Compound-Finance" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| borrow erc20 cToken | 0.00512864 | 0.007789121948713601 | 17 | 1 | 4154304 | 2564320 | 61.73 | 
| borrow ether cToken | 0.002804 | 0.0042585749719600005 | 14 | 1 | 2601104 | 1402000 | 53.9 | 
| Repay borrow erc20 cToken | 2e-05 | 3.0374999800000003e-05 | 18 | 1 | 1362240 | 10000 | 0.73 | 
| Mint erc20 cToken | 2e-05 | 3.0374999800000003e-05 | 17 | 1 | 1087592 | 10000 | 0.92 | 
| Mint ether cToken | 2e-05 | 3.0374999800000003e-05 | 12 | 1 | 296648 | 10000 | 3.37 | 
| Transfer erc20 cToken | 0.0116432 | 0.017683109883568 | 22 | 1 | 6788408 | 5821600 | 85.76 | 
| Add reserves for ether cToken | 2e-05 | 3.0374999800000003e-05 | 10 | 1 | 27500 | 10000 | 36.36 | 
| Liquidate borrow | 2e-05 | 3.0374999800000003e-05 | 21 | 1 | 1933536 | 10000 | 0.52 | 
| Redeem erc20 cToken | 2e-05 | 3.0374999800000003e-05 | 17 | 1 | 823944 | 10000 | 1.21 | 

Cost report for "Robonomics" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| PublicAmbix setSink | 0.00466928 | 0.0070914689533072 | 11 | 1 | 199023640 | 2334640 | 1.17 | 
| PublicAmbix appendSource | 0.00931856 | 0.014152562906814402 | 13 | 1 | 199023640 | 4659280 | 2.34 | 
| XRT approve | 0.00234464 | 0.0035609219765536 | 10 | 1 | 199022872 | 1172320 | 0.59 | 
| PublicAmbix run | 2e-05 | 3.0374999800000003e-05 | 21 | 1 | 199021400 | 10000 | 0.01 | 
| KycAmbix setSink | 0.00466928 | 0.0070914689533072 | 11 | 1 | 199023576 | 2334640 | 1.17 | 
| KycAmbix appendSource | 0.00931856 | 0.014152562906814402 | 13 | 1 | 199023576 | 4659280 | 2.34 | 
| KycAmbix run | 2e-05 | 3.0374999800000003e-05 | 20 | 1 | 199021400 | 10000 | 0.01 | 
| DutchAuction started | 2e-05 | 3.0374999800000003e-05 | 9 | 1 | 199021272 | 10000 | 0.01 | 
| DutchAuction bid | 0.00234464 | 0.0035609219765536 | 14 | 1 | 199026136 | 1172320 | 0.59 | 
| DutchAuction update stage | 2e-05 | 3.0391650000000002e-05 | 9 | 1 | 199021272 | 10000 | 0.01 | 
| DutchAuction claim tokens | 2e-05 | 3.0391650000000002e-05 | 13 | 1 | 199021272 | 10000 | 0.01 | 
| XRT transfer | 0.00234464 | 0.0035628739128 | 11 | 1 | 199023192 | 1172320 | 0.59 | 
| Factory create lighthouse | 0.0417104 | 0.063382393908 | 19 | 1 | 199022488 | 20855200 | 10.48 | 
| XRT increase allowance | 0.00234464 | 0.0035628680511999994 | 10 | 1 | 199022936 | 1172320 | 0.59 | 

Cost report for "Yearn" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Create new release | 0.00234464 | 0.0035628680511999994 | 14 | 1 | 1322552 | 1172320 | 88.64 | 
| Clone strategy | 0.03887072 | 0.059068917879999995 | 16 | 1 | 21455896 | 19435360 | 90.58 | 
| Set strategist authority | 2e-05 | 3.0392500000000004e-05 | 9 | 1 | 27500 | 10000 | 36.36 | 
| Set emergency shutdown | 2e-05 | 3.0391600000000002e-05 | 9 | 1 | 27500 | 10000 | 36.36 | 
| Emergency shutdown | 2e-05 | 3.0391600000000002e-05 | 9 | 1 | 27500 | 10000 | 36.36 | 
| Call strategy harvest | 0.00699392 | 0.010627820953599999 | 23 | 1 | 7854000 | 3496960 | 44.52 | 
| Withdraw from vault | 2e-05 | 3.0391600000000002e-05 | 15 | 1 | 823944 | 10000 | 1.21 | 

Cost report for "Uniswap V3" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Token approve | 0.00234464 | 0.0035628680511999994 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| NonfungiblePositionManager - Mint position | 0.02786064 | 0.0423364713312 | 4 | 28 | 31435120 | 13930320 | 44.31 | 
| NonfungiblePositionManager - Increase liquidity  | 0.00044 | 0.000668590472 | 23 | 23 | 4364240 | 220000 | 5.04 | 
| NonfungiblePositionManager - Decrease Liquidity | 0.00139808 | 0.0021241656199264 | 18 | 1 | 4274240 | 699040 | 16.35 | 
| NonfungiblePositionManager - Collect Fees | 0.00466928 | 0.007093312992057599 | 21 | 1 | 3553040 | 2334640 | 65.71 | 
| NonfungiblePositionManager - Burn Liquidity Position | 2e-05 | 3.03828984e-05 | 16 | 1 | 3954560 | 10000 | 0.25 | 
| Direct swap | 2e-05 | 3.03828984e-05 | 18 | 1 | 1328400 | 10000 | 0.75 | 
| Burn transaction | 0.00047936 | 0.0007282173088512 | 12 | 1 | 739040 | 239680 | 32.43 | 
| Collect transaction | 0.00234464 | 0.0035618479452287997 | 13 | 1 | 1661680 | 1172320 | 70.55 | 

Cost report for "Saddle Finance" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Add liquidity | 0.00235464 | 0.0035770393944287997 | 3 | 2 | 3209760 | 1177320 | 36.68 | 
| Swap DAI -> USDC | 2e-05 | 3.03828984e-05 | 22 | 1 | 1797760 | 10000 | 0.56 | 
| Swap USDC -> DAI | 2e-05 | 3.03828984e-05 | 22 | 1 | 1797760 | 10000 | 0.56 | 
| Remove liquidity | 2e-05 | 3.03828984e-05 | 24 | 1 | 1937440 | 10000 | 0.52 | 
| Add liquidity in 3 Tokens | 0.00284464 | 0.0043214204052288 | 4 | 31 | 4248480 | 1422320 | 33.48 | 
| Swap DAI -> USDC 3 pool | 2e-05 | 3.03828984e-05 | 22 | 1 | 1807760 | 10000 | 0.55 | 
| Swap USDC -> USDT 3 pool | 2e-05 | 3.03828984e-05 | 22 | 1 | 1807760 | 10000 | 0.55 | 
| Swap USDC -> DAI 3 pool | 2e-05 | 3.03828984e-05 | 22 | 1 | 1807760 | 10000 | 0.55 | 
| Remove liquidity 3 pool | 3e-05 | 4.55743476e-05 | 3 | 2 | 2646480 | 15000 | 0.57 | 
| Add liquidity in metapool | 0.00291464 | 0.0044277605496288 | 3 | 31 | 3299760 | 1457320 | 44.16 | 
| Swap SUSD -> LP metapool | 2e-05 | 3.03828984e-05 | 23 | 1 | 1807760 | 10000 | 0.55 | 
| Swap USDC -> SUSD metapool | 0.00103 | 0.0015647192676 | 4 | 53 | 3775520 | 515000 | 13.64 | 
| Swap DAI -> USDT metapool | 0.00056 | 0.0008507211551999999 | 4 | 30 | 3086160 | 280000 | 9.07 | 
| Remove liquidity in Metapool | 2e-05 | 3.03828984e-05 | 24 | 1 | 1937440 | 10000 | 0.52 | 

Cost report for "Uniswap V2" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Token approve | 0.00234464 | 0.0035618479452287997 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| Direct swap | 2e-05 | 3.03828984e-05 | 19 | 1 | 1048720 | 10000 | 0.95 | 

Cost report for "Curve" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Metapool: add liquidity | 0.01664256 | 0.0252824604797952 | 22 | 229 | 12479456 | 8321280 | 66.68 | 
| Metapool: swap 1 | 0.00458 | 0.006957683733599999 | 19 | 230 | 4803832 | 2290000 | 47.67 | 
| Metapool: swap 2 | 0.00464 | 0.0070488324288 | 19 | 233 | 4836832 | 2320000 | 47.97 | 
| Metapool: withdraw admin fees | 2e-05 | 3.03828984e-05 | 16 | 1 | 549296 | 10000 | 1.82 | 
| Pool: add liquidity | 0.01384464 | 0.0210320145252288 | 26 | 577 | 11162624 | 6922320 | 62.01 | 
| Pool: swap 1 | 0.00542 | 0.0082337654664 | 19 | 272 | 5265832 | 2710000 | 51.46 | 
| Pool: swap 2 | 0.00542 | 0.0082337654664 | 19 | 272 | 5265832 | 2710000 | 51.46 | 
| Pool: swap 3 | 0.00542 | 0.0082337654664 | 19 | 272 | 5265832 | 2710000 | 51.46 | 
| Pool: withdraw admin fees | 2e-05 | 3.03828984e-05 | 17 | 1 | 55000 | 10000 | 18.18 | 
| Stablecoin: transfer | 2e-05 | 3.03828984e-05 | 11 | 1 | 538296 | 10000 | 1.86 | 
| Tri pool: add liquidity | 0.01561656 | 0.0237238177918752 | 3 | 39 | 13440856 | 7808280 | 58.09 | 
| Tri pool: swap 1 | 0.00072 | 0.0010937843424 | 20 | 37 | 2680832 | 360000 | 13.43 | 
| Tri pool: swap 2 | 0.00072 | 0.0010937843424 | 20 | 37 | 2680832 | 360000 | 13.43 | 
| Tri pool: swap 3 | 0.0007 | 0.001063401444 | 20 | 36 | 2669832 | 350000 | 13.11 | 
| Tri pool: apply new fee | 2e-05 | 3.03828984e-05 | 9 | 1 | 27500 | 10000 | 36.36 | 
| Tri pool: withdraw admin fees | 2e-05 | 3.03828984e-05 | 19 | 1 | 560296 | 10000 | 1.78 | 
| gauge v3: deposit | 0.01907648 | 0.0289799376834816 | 29 | 1 | 11085360 | 9538240 | 86.04 | 
| gauge controller: vote | 0.01812992 | 0.027541975868006403 | 27 | 1 | 14547720 | 9064960 | 62.31 | 
| gauge controller: vote | 0.01069664 | 0.0162497463170688 | 27 | 1 | 10444104 | 5348320 | 51.21 | 
| gauge controller: vote | 0.01069664 | 0.0162497463170688 | 27 | 1 | 10444104 | 5348320 | 51.21 | 
| gauge controller: vote | 0.00791264 | 0.0120204468597888 | 27 | 1 | 8392296 | 3956320 | 47.14 | 
| gauge controller: vote | 0.005588 | 0.008488981812960001 | 27 | 1 | 7366392 | 2794000 | 37.93 | 
| gauge controller: vote | 0.005588 | 0.008488981812960001 | 27 | 1 | 7366392 | 2794000 | 37.93 | 
| gauge controller: vote | 0.00791264 | 0.0120204468597888 | 27 | 1 | 8392296 | 3956320 | 47.14 | 
| gauge controller: vote | 0.005588 | 0.008488981812960001 | 27 | 1 | 7366392 | 2794000 | 37.93 | 
| gauge controller: vote | 0.005588 | 0.008488981812960001 | 27 | 1 | 7366392 | 2794000 | 37.93 |

andreisilviudragnea commented on evm_loader/lib/src/commands/emulate.rs:
```diff
@@ -97,18 +99,22 @@ async fn emulate_trx<T: Tracer>(
 
-    let actions = backend.into_actions();
+        (backend.into_actions(), exit_status, steps_executed, tracer)
+    };
```
I suggest only reducing the scope of `let mut evm` and the mutable borrow needed by it in `evm.execute(step_limit, &mut backend).await?`, so the block can be even shorter:
```Rust
    let (exit_status, steps_executed, tracer) = {
        let mut evm = match Machine::new(&tx, origin, &mut backend, tracer).await {
            Ok(evm) => evm,
            Err(e) => return Ok((EmulateResponse::revert(e), None)),
        };

        evm.execute(step_limit, &mut backend).await?
    };
```

evgenyzdanovich replied:
thanks, done.

andreisilviudragnea commented on evm_loader/program/src/types/mod.rs:
```diff
@@ -4,7 +4,13 @@ pub use transaction::LegacyTx;
+pub use vector::Vector;
 
 mod address;
 pub mod serde;
```
The `pub mod serde;` can be removed completely, it is never used.

evgenyzdanovich replied:
done

andreisilviudragnea commented on evm_loader/program/Cargo.toml:
```diff
@@ -56,6 +56,7 @@ cfg-if = { version = "1.0" }
 log = { version = "0.4", default-features = false, optional = true }
 maybe-async = "0.2.10"
 async-trait = { version = "0.1.79", optional = true }
+allocator-api2 = "0.2.16"
```
In this file:
- the `serde_bytes` dependency can be removed;
- the `[dev-dependencies]` section can be removed completely, since those dependencies are no longer used.

evgenyzdanovich replied:
done

andreisilviudragnea commented on evm_loader/program/src/evm/memory.rs:
```diff
@@ -14,6 +14,7 @@ const MEMORY_ALIGN: usize = 1;
 static_assertions::const_assert!(MEMORY_ALIGN.is_power_of_two());
 
+#[repr(C)]
 pub struct Memory {
```
In this file, `serde::Serialize` and `serde::Deserialize` impls can be removed.

evgenyzdanovich replied:
done

andreisilviudragnea commented on evm_loader/program/src/evm/stack.rs:
```diff
@@ -12,6 +12,7 @@ use crate::{error::Error, types::Address};
 const STACK_SIZE: usize = ELEMENT_SIZE * 128;
 
+#[repr(C)]
 pub struct Stack {
```
In this file, `serde::Serialize` and `serde::Deserialize` impls can be removed.

evgenyzdanovich replied:
done

andreisilviudragnea reviewed (commented):
This is my first round of comments, I am still reviewing the PR.

evgenyzdanovich committed:
Code review, part 3.

neonlabstech commented:
<details><summary>Dapps report</summary>


Cost report for "Aave" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Token mint | 0.00234464 | 0.0033933974719999997 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| Token approve | 0.00234464 | 0.0033933974719999997 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| Deposit to lending pool | 0.00796264 | 0.011524328872 | 4 | 5 | 4845360 | 3981320 | 82.17 | 
| Borrow from lending pool | 0.00519928 | 0.007524917944 | 4 | 30 | 3793040 | 2599640 | 68.54 | 
| Repay | 0.00049 | 0.000709177 | 4 | 27 | 2586800 | 245000 | 9.47 | 
| Flashloan | 0.00512928 | 0.0074238184755072 | 4 | 29 | 7499680 | 2564640 | 34.2 | 
| Withdraw | 7e-05 | 0.0001013138868 | 4 | 5 | 1857760 | 35000 | 1.88 | 
| Liquidation | 0.00341464 | 0.004942150691756 | 4 | 57 | 4967840 | 1707320 | 34.37 | 

Cost report for "Compound-Finance" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| borrow erc20 cToken | 0.00512864 | 0.0074260661385504015 | 17 | 1 | 4154304 | 2564320 | 61.73 | 
| borrow ether cToken | 0.002804 | 0.0040600801484400005 | 14 | 1 | 2601104 | 1402000 | 53.9 | 
| Repay borrow erc20 cToken | 2e-05 | 2.8960999800000006e-05 | 18 | 1 | 1362240 | 10000 | 0.73 | 
| Mint erc20 cToken | 2e-05 | 2.8961158600000005e-05 | 17 | 1 | 1087592 | 10000 | 0.92 | 
| Mint ether cToken | 2e-05 | 2.8961158600000005e-05 | 12 | 1 | 296648 | 10000 | 3.37 | 
| Transfer erc20 cToken | 0.0116432 | 0.016860028090576 | 22 | 1 | 6788408 | 5821600 | 85.76 | 
| Add reserves for ether cToken | 2e-05 | 2.8961158600000005e-05 | 10 | 1 | 27500 | 10000 | 36.36 | 
| Liquidate borrow | 2e-05 | 2.8961158600000005e-05 | 21 | 1 | 1933536 | 10000 | 0.52 | 
| Redeem erc20 cToken | 2e-05 | 2.89610106e-05 | 17 | 1 | 823944 | 10000 | 1.21 | 

Cost report for "Robonomics" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| PublicAmbix setSink | 0.00466928 | 0.006761664726308799 | 11 | 1 | 199023640 | 2334640 | 1.17 | 
| PublicAmbix appendSource | 0.00931856 | 0.0134943671084176 | 13 | 1 | 199023640 | 4659280 | 2.34 | 
| XRT approve | 0.00234464 | 0.003395390416 | 10 | 1 | 199022872 | 1172320 | 0.59 | 
| PublicAmbix run | 2e-05 | 2.8963000000000004e-05 | 21 | 1 | 199021400 | 10000 | 0.01 | 
| KycAmbix setSink | 0.00466928 | 0.006761213300318399 | 11 | 1 | 199023576 | 2334640 | 1.17 | 
| KycAmbix appendSource | 0.00931856 | 0.0134934661900368 | 13 | 1 | 199023576 | 4659280 | 2.34 | 
| KycAmbix run | 2e-05 | 2.8960410600000002e-05 | 20 | 1 | 199021400 | 10000 | 0.01 | 
| DutchAuction started | 2e-05 | 2.8960410600000002e-05 | 9 | 1 | 199021272 | 10000 | 0.01 | 
| DutchAuction bid | 0.00234464 | 0.0033950868554592 | 14 | 1 | 199026200 | 1172320 | 0.59 | 
| DutchAuction update stage | 2e-05 | 2.8960410600000002e-05 | 9 | 1 | 199021272 | 10000 | 0.01 | 
| DutchAuction claim tokens | 2e-05 | 2.8963000000000004e-05 | 13 | 1 | 199021272 | 10000 | 0.01 | 
| XRT transfer | 0.00234464 | 0.0033954090089952 | 11 | 1 | 199023192 | 1172320 | 0.59 | 
| Factory create lighthouse | 0.0417104 | 0.060403634013088 | 19 | 1 | 199022488 | 20855200 | 10.48 | 
| XRT increase allowance | 0.00234464 | 0.0033954307907007996 | 10 | 1 | 199022936 | 1172320 | 0.59 | 

Cost report for "Yearn" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Create new release | 0.00234464 | 0.0033954307907007996 | 14 | 1 | 1322552 | 1172320 | 88.64 | 
| Clone strategy | 0.03887072 | 0.0562915322477536 | 16 | 1 | 21455896 | 19435360 | 90.58 | 
| Set strategist authority | 2e-05 | 2.8963158600000002e-05 | 9 | 1 | 27500 | 10000 | 36.36 | 
| Set emergency shutdown | 2e-05 | 2.8963158600000002e-05 | 9 | 1 | 27500 | 10000 | 36.36 | 
| Emergency shutdown | 2e-05 | 2.8963000000000004e-05 | 9 | 1 | 27500 | 10000 | 36.36 | 
| Call strategy harvest | 0.00699392 | 0.010127758331289599 | 23 | 1 | 7854000 | 3496960 | 44.52 | 
| Withdraw from vault | 2e-05 | 2.8965108000000003e-05 | 15 | 1 | 823944 | 10000 | 1.21 | 

Cost report for "Uniswap V3" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Token approve | 0.00234464 | 0.0033958874796799995 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| NonfungiblePositionManager - Mint position | 0.02786064 | 0.040353683632041594 | 4 | 28 | 31435120 | 13930320 | 44.31 | 
| NonfungiblePositionManager - Increase liquidity  | 0.00044 | 0.0006373012536 | 23 | 23 | 4364240 | 220000 | 5.04 | 
| NonfungiblePositionManager - Decrease Liquidity | 0.00139808 | 0.0020249957650751997 | 18 | 1 | 4274240 | 699040 | 16.35 | 
| NonfungiblePositionManager - Collect Fees | 0.00466928 | 0.0067632992544656 | 21 | 1 | 3553040 | 2334640 | 65.71 | 
| NonfungiblePositionManager - Burn Liquidity Position | 2e-05 | 2.8970200000000003e-05 | 16 | 1 | 3954560 | 10000 | 0.25 | 
| Direct swap | 2e-05 | 2.8969345400000004e-05 | 18 | 1 | 1328400 | 10000 | 0.75 | 
| Burn transaction | 0.00047936 | 0.0006943372705472 | 12 | 1 | 739040 | 239680 | 32.43 | 
| Collect transaction | 0.00234464 | 0.0033957707635007997 | 13 | 1 | 1661680 | 1172320 | 70.55 | 

Cost report for "Saddle Finance" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Add liquidity | 0.00235464 | 0.0034102538857008 | 3 | 2 | 3209760 | 1177320 | 36.68 | 
| Swap DAI -> USDC | 2e-05 | 2.8966244400000004e-05 | 22 | 1 | 1797760 | 10000 | 0.56 | 
| Swap USDC -> DAI | 2e-05 | 2.8966244400000004e-05 | 22 | 1 | 1797760 | 10000 | 0.56 | 
| Remove liquidity | 2e-05 | 2.8967952200000003e-05 | 24 | 1 | 1937440 | 10000 | 0.52 | 
| Add liquidity in 3 Tokens | 0.00284464 | 0.0041204538715072 | 4 | 31 | 4248480 | 1422320 | 33.48 | 
| Swap DAI -> USDC 3 pool | 2e-05 | 2.8968509000000004e-05 | 22 | 1 | 1807760 | 10000 | 0.55 | 
| Swap USDC -> USDT 3 pool | 2e-05 | 2.8970345400000002e-05 | 22 | 1 | 1807760 | 10000 | 0.55 | 
| Swap USDC -> DAI 3 pool | 2e-05 | 2.8970345400000002e-05 | 22 | 1 | 1807760 | 10000 | 0.55 | 
| Remove liquidity 3 pool | 3e-05 | 4.34555181e-05 | 3 | 2 | 2646480 | 15000 | 0.57 | 
| Add liquidity in metapool | 0.00291464 | 0.0042219063758328 | 3 | 31 | 3299760 | 1457320 | 44.16 | 
| Swap SUSD -> LP metapool | 2e-05 | 2.8971958e-05 | 23 | 1 | 1807760 | 10000 | 0.55 | 
| Swap USDC -> SUSD metapool | 0.00103 | 0.0014920582060000001 | 4 | 53 | 3775520 | 515000 | 13.64 | 
| Swap DAI -> USDT metapool | 0.00056 | 0.0008112161119999999 | 4 | 30 | 3086160 | 280000 | 9.07 | 
| Remove liquidity in Metapool | 2e-05 | 2.8973150800000005e-05 | 24 | 1 | 1937440 | 10000 | 0.52 | 

Cost report for "Uniswap V2" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Token approve | 0.00234464 | 0.0033965804145856 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| Direct swap | 2e-05 | 2.89722e-05 | 19 | 1 | 1048720 | 10000 | 0.95 | 

Cost report for "Curve" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Metapool: add liquidity | 0.01664256 | 0.0241085788416 | 22 | 229 | 12479456 | 8321280 | 66.68 | 
| Metapool: swap 1 | 0.00458 | 0.006634633799999999 | 19 | 230 | 4803832 | 2290000 | 47.67 | 
| Metapool: swap 2 | 0.00464 | 0.0067214942560000004 | 19 | 233 | 4836832 | 2320000 | 47.97 | 
| Metapool: withdraw admin fees | 2e-05 | 2.8972011200000003e-05 | 16 | 1 | 549296 | 10000 | 1.82 | 
| Pool: add liquidity | 0.01384464 | 0.020056740489926398 | 26 | 577 | 11162624 | 6922320 | 62.01 | 
| Pool: swap 1 | 0.00542 | 0.0078519581192 | 19 | 272 | 5265832 | 2710000 | 51.46 | 
| Pool: swap 2 | 0.00542 | 0.0078519581192 | 19 | 272 | 5265832 | 2710000 | 51.46 | 
| Pool: swap 3 | 0.00542 | 0.0078519581192 | 19 | 272 | 5265832 | 2710000 | 51.46 | 
| Pool: withdraw admin fees | 2e-05 | 2.89740152e-05 | 17 | 1 | 55000 | 10000 | 18.18 | 
| Stablecoin: transfer | 2e-05 | 2.89740152e-05 | 11 | 1 | 538296 | 10000 | 1.86 | 
| Tri pool: add liquidity | 0.01561656 | 0.022623722340585597 | 3 | 39 | 13440856 | 7808280 | 58.09 | 
| Tri pool: swap 1 | 0.00072 | 0.001043136 | 20 | 37 | 2680832 | 360000 | 13.43 | 
| Tri pool: swap 2 | 0.00072 | 0.001043136 | 20 | 37 | 2680832 | 360000 | 13.43 | 
| Tri pool: swap 3 | 0.0007 | 0.00101416 | 20 | 36 | 2669832 | 350000 | 13.11 | 
| Tri pool: apply new fee | 2e-05 | 2.8976000000000005e-05 | 9 | 1 | 27500 | 10000 | 36.36 | 
| Tri pool: withdraw admin fees | 2e-05 | 2.8976000000000005e-05 | 19 | 1 | 560296 | 10000 | 1.78 | 
| gauge v3: deposit | 0.01907648 | 0.027638004224 | 29 | 1 | 11085360 | 9538240 | 86.04 | 
| gauge controller: vote | 0.01812992 | 0.026266628096000003 | 27 | 1 | 14547720 | 9064960 | 62.31 | 
| gauge controller: vote | 0.01069664 | 0.015497292032000001 | 27 | 1 | 10444104 | 5348320 | 51.21 | 
| gauge controller: vote | 0.01069664 | 0.015497292032000001 | 27 | 1 | 10444104 | 5348320 | 51.21 | 
| gauge controller: vote | 0.00791264 | 0.011463832832 | 27 | 1 | 8392296 | 3956320 | 47.14 | 
| gauge controller: vote | 0.005588 | 0.008095894400000001 | 27 | 1 | 7366392 | 2794000 | 37.93 | 
| gauge controller: vote | 0.005588 | 0.008095894400000001 | 27 | 1 | 7366392 | 2794000 | 37.93 | 
| gauge controller: vote | 0.00791264 | 0.011463832832 | 27 | 1 | 8392296 | 3956320 | 47.14 | 
| gauge controller: vote | 0.005588 | 0.008095894400000001 | 27 | 1 | 7366392 | 2794000 | 37.93 | 
| gauge controller: vote | 0.005588 | 0.008095894400000001 | 27 | 1 | 7366392 | 2794000 | 37.93 |

andreisilviudragnea committed:
Refactor vector usages

Replace vect fn with vector! macro

Replace into_vector(value.to_vec()) with value.to_vector()

Convert into_vector function to method

andreisilviudragnea committed:
Refactor Buffer usages

Optimize Buffer::from_slice() to convert from slice to Vector directly

Replace Buffer::from_vec with Buffer::from_vector

andreisilviudragnea committed:
Replace BUFFER_OFFSET with MIN_HEAP_OFFSET

andreisilviudragnea commented on evm_loader/program/src/types/mod.rs:
```diff
@@ -4,7 +4,12 @@ pub use transaction::LegacyTx;
+pub use vector::Vector;
 
 mod address;
-pub mod serde;
```
There is also a file that needs to be deleted: `evm_loader/program/src/types/serde.rs`

andreisilviudragnea commented on evm_loader/program/src/types/vector.rs:
```diff
@@ -0,0 +1,34 @@
+#[macro_export]
+macro_rules! vector {
+    () => (
+        allocator_api2::vec::Vec::with_capacity_in(1, $crate::allocator::acc_allocator())
```
I suggest replacing this with:
```Rust
        allocator_api2::vec::Vec::new_in($crate::allocator::acc_allocator())
```

andreisilviudragnea commented on evm_loader/program/src/types/vector.rs:
```diff
@@ -0,0 +1,34 @@
+#[must_use]
+pub fn vect<T>() -> Vector<T> {
+    Vector::with_capacity_in(1, acc_allocator())
+}
```
The `vect()` function can be replaced by usages of `vector![]` macro call.

andreisilviudragnea commented on evm_loader/program/src/executor/precompile_extension/spl_token.rs:
```diff
@@ -578,10 +587,10 @@ impl<B: AccountStorage> ExecutorState<'_, B> {
-            vec![bump_seed],
+        let seeds = vector![
+            vector![ACCOUNT_SEED_VERSION],
+            into_vector(signer.as_bytes().to_vec()),
```
Usages like these:
- create a `Vec<T>` from a slice (`[T]`);
- then immediately create a `Vector<T>` from `Vec<T>`.

I suggest adding an extension method `to_vector()` to `[T]` (in `vector.rs` file) and use it to convert from a slice directly to a `Vector<T>`:
```Rust
use allocator_api2::SliceExt;

pub trait VectorSliceExt<T> {
    fn to_vector(&self) -> Vector<T>
    where
        T: Clone;
}

impl<T> VectorSliceExt<T> for [T] {
    fn to_vector(&self) -> Vector<T>
    where
        T: Clone,
    {
        SliceExt::to_vec_in(self, acc_allocator())
    }
}
```

This way, the above line becomes:
```Rust
signer.as_bytes().to_vector()
```
The same comment applies to all usages of the form `into_vector(slice.to_vec())` in this PR.

andreisilviudragnea commented on evm_loader/program/src/types/vector.rs:
```diff
@@ -0,0 +1,34 @@
+        ret.push(item);
+    }
+    ret
+}
```
For consistency with the suggested `to_vector()` method on slices (`[T]`), I suggest converting this function to an extension method on `Vec<T>`:
```Rust
pub trait VectorVecExt<T> {
    fn into_vector(self) -> Vector<T>;
}

impl<T> VectorVecExt<T> for Vec<T> {
    fn into_vector(self) -> Vector<T> {
        let mut ret = Vector::with_capacity_in(self.len(), acc_allocator());
        for item in self {
            ret.push(item);
        }
        ret
    }
}
```

andreisilviudragnea commented on evm_loader/program/src/evm/buffer.rs:
```diff
@@ -69,7 +69,7 @@ impl Buffer {
+        Self::new(Inner::Owned(into_vector(v)))
     }
 
     #[must_use]
```
I suggest adding a new method:
```Rust
    #[must_use]
    pub fn from_vector(data: Vector<u8>) -> Self {
        Self::new(Inner::Owned(data))
    }
```
and use it to optimize the implementation of `Buffer::from_slice`:
```Rust
    #[must_use]
    pub fn from_slice(v: &[u8]) -> Self {
        Self::from_vector(v.to_vector())
    }
```
to create a `Vector<T>` directly from a slice (`[T]`) instead of creating an intermediate `Vec<T>`.

The new `Buffer::from_vector` method can be used to simplify:
```Rust
    #[must_use]
    pub fn empty() -> Self {
        Self::from_vector(vector![])
    }
```
and also to replace:
```Rust
    impl Buffer {
        fn with_persistent_vector(data: Vector<u8>) -> Self {
            Self::new(Inner::Owned(data))
        }
    }
```

andreisilviudragnea commented on evm_loader/program/src/evm/opcode.rs:
```diff
@@ -1444,6 +1451,10 @@ impl<B: Database, T: EventListener> Machine<B, T> {
 
         self.return_data = Buffer::from_vec(return_data);
```
`self.return_data = Buffer::from_vec(return_data);` is assigned only in `opcode_return_impl` and `opcode_revert_impl` methods.

The input `return_data` is of type `Vec<u8>`, which is then moved into a `Vector<u8>` inside a `Buffer`, because the destination of `return_data` is always `self.return_data: Buffer` field of `Machine`.

I suggest directly allocating and returning `Vector<u8>` from all places which return data for this field and then finally replacing the line:
```Rust
Buffer::from_vec(return_data);
```
with
```Rust
Buffer::from_vector(return_data);
```

The purpose of this suggestion is to finally remove the `Buffer::from_vec` method which forces conversions from `Vec<u8>` to `Vector<u8>`, instead of using `Vector<u8>` directly.

andreisilviudragnea commented on evm_loader/program/src/evm/buffer.rs:
```diff
@@ -69,7 +69,7 @@ impl Buffer {
     pub fn from_vec(v: Vec<u8>) -> Self {
-        Self::new(Inner::Owned(v))
+        Self::new(Inner::Owned(into_vector(v)))
     }
```
Taking into account my other comments, there should be only two usages of this `Buffer::from_vec` method inside `EmulatorAccountStorage::code` method, which can be replaced with `Buffer::from_vector` and explicit `.into_vector()` calls, and then this method can be removed:
```Rust
    async fn code(&self, address: Address) -> evm_loader::evm::Buffer {
        use evm_loader::evm::Buffer;

        info!("code {address}");

        let code_override = self.account_override(address, |a| a.code.clone());
        if let Some(code_override) = code_override {
            return Buffer::from_vector(code_override.0.into_vector());
        }

        let code = self
            .ethereum_contract_map_or(
                address,
                Vec::default(),
                |legacy, info| legacy.read_code(info),
                |c| c.code().to_vec(),
            )
            .await;

        Buffer::from_vector(code.into_vector())
    }
```
I believe that having more conversions like these explicit is better and they also happen in `lib` crate, so outside of NeonEVM Solana program itself.

andreisilviudragnea commented on evm_loader/program/src/account/holder.rs:
```diff
@@ -180,6 +186,23 @@ impl<'a> Holder<'a> {
+    /// N.B. No ownership checks are performed, it's a caller's responsibility.
+    /// TODO: This piece of should be moved to mod.rs.
+    pub fn init_heap(&mut self, transaction_offset: usize) -> Result<()> {
+        let buffer_end = BUFFER_OFFSET + transaction_offset;
```
I suggest using `MIN_HEAP_OFFSET` here too instead of `BUFFER_OFFSET` just like when calling:
```Rust
let actual_heap_offset = init_heap(&accounts[0], MIN_HEAP_OFFSET);
```
from `transaction_step_from_instruction.rs`.

andreisilviudragnea reviewed (commented):
This is the second round of comments. I am still reviewing the PR. Let's discuss these comments after daily meeting.

evgenyzdanovich commented:
Resolving current conversations with Andrei - we discussed it by voice and I cherry-picked changes. 
Thanks for contributions, I highly appreciate it!

neonlabstech commented:
<details><summary>Dapps report</summary>


Cost report for "Aave" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Token mint | 0.00234464 | 0.0035598468653279996 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| Token approve | 0.00234464 | 0.0035598468653279996 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| Deposit to lending pool | 0.00796264 | 0.012089608231428 | 4 | 5 | 4845360 | 3981320 | 82.17 | 
| Borrow from lending pool | 0.00519928 | 0.007894022370156 | 4 | 30 | 3793040 | 2599640 | 68.54 | 
| Repay | 0.00049 | 0.0007439615855000001 | 4 | 27 | 2586800 | 245000 | 9.47 | 
| Flashloan | 0.00512928 | 0.007787923852924801 | 4 | 29 | 7499680 | 2564640 | 34.2 | 
| Withdraw | 7e-05 | 0.0001062802265 | 4 | 5 | 1857760 | 35000 | 1.88 | 
| Liquidation | 0.00341464 | 0.0051844101802280005 | 4 | 57 | 4967840 | 1707320 | 34.37 | 

Cost report for "Compound-Finance" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| borrow erc20 cToken | 0.00512864 | 0.007786757440528 | 17 | 1 | 4154304 | 2564320 | 61.73 | 
| borrow ether cToken | 0.002804 | 0.0042572822158000005 | 14 | 1 | 2601104 | 1402000 | 53.9 | 
| Repay borrow erc20 cToken | 2e-05 | 3.0368045000000004e-05 | 18 | 1 | 1362240 | 10000 | 0.73 | 
| Mint erc20 cToken | 2e-05 | 3.0368045000000004e-05 | 17 | 1 | 1087592 | 10000 | 0.92 | 
| Mint ether cToken | 2e-05 | 3.0368045000000004e-05 | 12 | 1 | 296648 | 10000 | 3.37 | 
| Transfer erc20 cToken | 0.0116432 | 0.0176790610772 | 22 | 1 | 6788408 | 5821600 | 85.76 | 
| Add reserves for ether cToken | 2e-05 | 3.0365779000000002e-05 | 10 | 1 | 27500 | 10000 | 36.36 | 
| Liquidate borrow | 2e-05 | 3.0362336000000004e-05 | 21 | 1 | 1933536 | 10000 | 0.52 | 
| Redeem erc20 cToken | 2e-05 | 3.0368045000000004e-05 | 17 | 1 | 823944 | 10000 | 1.21 | 

Cost report for "Robonomics" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| PublicAmbix setSink | 0.00466928 | 0.0070898452578799995 | 11 | 1 | 199023640 | 2334640 | 1.17 | 
| PublicAmbix appendSource | 0.00931856 | 0.0141479733296432 | 13 | 1 | 199023640 | 4659280 | 2.34 | 
| XRT approve | 0.00234464 | 0.0035597671944607996 | 10 | 1 | 199022872 | 1172320 | 0.59 | 
| PublicAmbix run | 2e-05 | 3.0364648200000002e-05 | 21 | 1 | 199021400 | 10000 | 0.01 | 
| KycAmbix setSink | 0.00466928 | 0.007089052227364799 | 11 | 1 | 199023576 | 2334640 | 1.17 | 
| KycAmbix appendSource | 0.00931856 | 0.0141477398065296 | 13 | 1 | 199023576 | 4659280 | 2.34 | 
| KycAmbix run | 2e-05 | 3.0364648200000002e-05 | 20 | 1 | 199021400 | 10000 | 0.01 | 
| DutchAuction started | 2e-05 | 3.0364648200000002e-05 | 9 | 1 | 199021272 | 10000 | 0.01 | 
| DutchAuction bid | 0.00234464 | 0.0035597084377823997 | 14 | 1 | 199026200 | 1172320 | 0.59 | 
| DutchAuction update stage | 2e-05 | 3.0364648200000002e-05 | 9 | 1 | 199021272 | 10000 | 0.01 | 
| DutchAuction claim tokens | 2e-05 | 3.0366489400000004e-05 | 13 | 1 | 199021272 | 10000 | 0.01 | 
| XRT transfer | 0.00234464 | 0.0035599242853408 | 11 | 1 | 199023192 | 1172320 | 0.59 | 
| Factory create lighthouse | 0.0417104 | 0.063329513879984 | 19 | 1 | 199022488 | 20855200 | 10.48 | 
| XRT increase allowance | 0.00234464 | 0.0035599014016543997 | 10 | 1 | 199022936 | 1172320 | 0.59 | 

Cost report for "Yearn" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Create new release | 0.00234464 | 0.0035599229020032 | 14 | 1 | 1322552 | 1172320 | 88.64 | 
| Clone strategy | 0.03887072 | 0.059017785391376004 | 16 | 1 | 21455896 | 19435360 | 90.58 | 
| Set strategist authority | 2e-05 | 3.0366244600000004e-05 | 9 | 1 | 27500 | 10000 | 36.36 | 
| Set emergency shutdown | 2e-05 | 3.0366244600000004e-05 | 9 | 1 | 27500 | 10000 | 36.36 | 
| Emergency shutdown | 2e-05 | 3.0365723000000003e-05 | 9 | 1 | 27500 | 10000 | 36.36 | 
| Call strategy harvest | 0.00699392 | 0.010618956509696 | 23 | 1 | 7854000 | 3496960 | 44.52 | 
| Withdraw from vault | 2e-05 | 3.0366251000000004e-05 | 15 | 1 | 823944 | 10000 | 1.21 | 

Cost report for "Uniswap V3" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Token approve | 0.00234464 | 0.0035598349780031997 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| NonfungiblePositionManager - Mint position | 0.02786064 | 0.0423004302500832 | 4 | 28 | 31435120 | 13930320 | 44.31 | 
| NonfungiblePositionManager - Increase liquidity  | 0.00044 | 0.0006680461304 | 23 | 23 | 4364240 | 220000 | 5.04 | 
| NonfungiblePositionManager - Decrease Liquidity | 0.00139808 | 0.0021227414098111998 | 18 | 1 | 4274240 | 699040 | 16.35 | 
| NonfungiblePositionManager - Collect Fees | 0.00466928 | 0.007089321224585599 | 21 | 1 | 3553040 | 2334640 | 65.71 | 
| NonfungiblePositionManager - Burn Liquidity Position | 2e-05 | 3.0366541400000003e-05 | 16 | 1 | 3954560 | 10000 | 0.25 | 
| Direct swap | 2e-05 | 3.0366541400000003e-05 | 18 | 1 | 1328400 | 10000 | 0.75 | 
| Burn transaction | 0.00047936 | 0.0007277817096256 | 12 | 1 | 739040 | 239680 | 32.43 | 
| Collect transaction | 0.00234464 | 0.0035597173474143998 | 13 | 1 | 1661680 | 1172320 | 70.55 | 

Cost report for "Saddle Finance" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Add liquidity | 0.00235464 | 0.0035748407964216 | 3 | 2 | 3209760 | 1177320 | 36.68 | 
| Swap DAI -> USDC | 2e-05 | 3.0364223800000002e-05 | 22 | 1 | 1797760 | 10000 | 0.56 | 
| Swap USDC -> DAI | 2e-05 | 3.0364223800000002e-05 | 22 | 1 | 1797760 | 10000 | 0.56 | 
| Remove liquidity | 2e-05 | 3.0364223800000002e-05 | 24 | 1 | 1937440 | 10000 | 0.52 | 
| Add liquidity in 3 Tokens | 0.00284464 | 0.0043187642795215996 | 4 | 31 | 4248480 | 1422320 | 33.48 | 
| Swap DAI -> USDC 3 pool | 2e-05 | 3.0364223800000002e-05 | 22 | 1 | 1807760 | 10000 | 0.55 | 
| Swap USDC -> USDT 3 pool | 2e-05 | 3.0364625800000005e-05 | 22 | 1 | 1807760 | 10000 | 0.55 | 
| Swap USDC -> DAI 3 pool | 2e-05 | 3.0364625800000005e-05 | 22 | 1 | 1807760 | 10000 | 0.55 | 
| Remove liquidity 3 pool | 3e-05 | 4.554693870000001e-05 | 3 | 2 | 2646480 | 15000 | 0.57 | 
| Add liquidity in metapool | 0.00291464 | 0.004425097647085601 | 3 | 31 | 3299760 | 1457320 | 44.16 | 
| Swap SUSD -> LP metapool | 2e-05 | 3.0364625800000005e-05 | 23 | 1 | 1807760 | 10000 | 0.55 | 
| Swap USDC -> SUSD metapool | 0.00103 | 0.0015637221864 | 4 | 53 | 3775520 | 515000 | 13.64 | 
| Swap DAI -> USDT metapool | 0.00056 | 0.0008499353239999999 | 4 | 30 | 3086160 | 280000 | 9.07 | 
| Remove liquidity in Metapool | 2e-05 | 3.0354833000000003e-05 | 24 | 1 | 1937440 | 10000 | 0.52 | 

Cost report for "Uniswap V2" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Token approve | 0.00234464 | 0.003558462589872 | 10 | 1 | 1192320 | 1172320 | 98.32 | 
| Direct swap | 2e-05 | 3.0354021000000005e-05 | 19 | 1 | 1048720 | 10000 | 0.95 | 

Cost report for "Curve" dApp

| Action | Fee | Cost in $ | Accounts | TRx | Estimated Gas | Used Gas | Used % of EG |
| --- | --- | --- | --- | --- | --- | --- |--- |
| Metapool: add liquidity | 0.01664256 | 0.025258430786688 | 22 | 229 | 12479456 | 8321280 | 66.68 | 
| Metapool: swap 1 | 0.00458 | 0.0069470666982 | 19 | 230 | 4803832 | 2290000 | 47.67 | 
| Metapool: swap 2 | 0.00464 | 0.0070388248304000004 | 19 | 233 | 4836832 | 2320000 | 47.97 | 
| Metapool: withdraw admin fees | 2e-05 | 3.0340000000000004e-05 | 16 | 1 | 549296 | 10000 | 1.82 | 
| Pool: add liquidity | 0.01384464 | 0.021002318880000002 | 26 | 577 | 11162624 | 6922320 | 62.01 | 
| Pool: swap 1 | 0.00542 | 0.0082220755562 | 19 | 272 | 5265832 | 2710000 | 51.46 | 
| Pool: swap 2 | 0.00542 | 0.0082220755562 | 19 | 272 | 5265832 | 2710000 | 51.46 | 
| Pool: swap 3 | 0.00542 | 0.008222439184 | 19 | 272 | 5265832 | 2710000 | 51.46 | 
| Pool: withdraw admin fees | 2e-05 | 3.0339762200000003e-05 | 17 | 1 | 55000 | 10000 | 18.18 | 
| Stablecoin: transfer | 2e-05 | 3.0339762200000003e-05 | 11 | 1 | 538296 | 10000 | 1.86 | 
| Tri pool: add liquidity | 0.01561656 | 0.0236901358391016 | 3 | 39 | 13440856 | 7808280 | 58.09 | 
| Tri pool: swap 1 | 0.00072 | 0.0010922314392 | 20 | 37 | 2680832 | 360000 | 13.43 | 
| Tri pool: swap 2 | 0.00072 | 0.0010923484032 | 20 | 37 | 2680832 | 360000 | 13.43 | 
| Tri pool: swap 3 | 0.0007 | 0.001062005392 | 20 | 36 | 2669832 | 350000 | 13.11 | 
| Tri pool: apply new fee | 2e-05 | 3.03452286e-05 | 9 | 1 | 27500 | 10000 | 36.36 | 
| Tri pool: withdraw admin fees | 2e-05 | 3.03452286e-05 | 19 | 1 | 560296 | 10000 | 1.78 | 
| gauge v3: deposit | 0.01907648 | 0.0289440073241664 | 29 | 1 | 11085360 | 9538240 | 86.04 | 
| gauge controller: vote | 0.01812992 | 0.0275078283449856 | 27 | 1 | 14547720 | 9064960 | 62.31 | 
| gauge controller: vote | 0.01069664 | 0.0162295993025952 | 27 | 1 | 10444104 | 5348320 | 51.21 | 
| gauge controller: vote | 0.01069664 | 0.0162295993025952 | 27 | 1 | 10444104 | 5348320 | 51.21 | 
| gauge controller: vote | 0.00791264 | 0.0120048748633952 | 27 | 1 | 8392296 | 3956320 | 47.14 | 
| gauge controller: vote | 0.005588 | 0.00847798468484 | 27 | 1 | 7366392 | 2794000 | 37.93 | 
| gauge controller: vote | 0.005588 | 0.00847798468484 | 27 | 1 | 7366392 | 2794000 | 37.93 | 
| gauge controller: vote | 0.00791264 | 0.0120048748633952 | 27 | 1 | 8392296 | 3956320 | 47.14 | 
| gauge controller: vote | 0.005588 | 0.00847798468484 | 27 | 1 | 7366392 | 2794000 | 37.93 | 
| gauge controller: vote | 0.005588 | 0.00847822077784 | 27 | 1 | 7366392 | 2794000 | 37.93 |
